File: leetcode/704e.py
"""
Given an array of integers nums which is sorted in ascending order, and an integer target, write a function to search target in nums. If target exists, then return its index. Otherwise, return -1.

You must write an algorithm with O(log n) runtime complexity.
"""
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while left <= right:
            middle = (left + right) // 2
            if nums[middle] == target:
                return middle
            elif left == right:
                return -1
            if nums[middle] > target:
                right = middle - 1
            else:
                left = middle + 1
        return -1

File: leetcode/704e.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        int middle = -1;
        int num;

        while (left <= right) {
            middle = (left + right) / 2;
            num = nums[middle];
            if (num == target) {
                return middle;
            }
            if (num < target) {
                left = middle + 1;
            } else if (num > target) {
                right = middle - 1;
            }
        }
        return -1;
    }
}
